src/main/java/com/google/finapp/SpannerDao.java

```diff
@@ -16,133 +16,183 @@
+        ps.setBytes(1, accountId.toByteArray());
+        ps.setInt(2, accountType.getNumber());
+        ps.setInt(3, accountStatus.getNumber());
+        ps.setBigDecimal(4, balance);
+        ps.executeUpdate();
+        System.out.printf("Account created with balance %s\n", balance.toString());
+      }
+    }
   }
 
   void addAccountForCustomer(
       ByteArray customerId, ByteArray accountId, ByteArray roleId, String roleName)
-      throws SpannerException {
-    databaseClient.write(
-        ImmutableList.of(
-            Mutation.newInsertBuilder("CustomerRole")
-                .set("CustomerId")
-                .to(customerId)
-                .set("AccountId")
-                .to(accountId)
-                .set("RoleId")
-                .to(roleId)
-                .set("Role")
-                .to(roleName)
-                .build()));
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      try (PreparedStatement ps =
+          connection.prepareStatement(
+              "INSERT INTO CustomerRole\n"
+                  + "(CustomerId, AccountId, RoleId, Role)\n"
+                  + "VALUES\n"
+                  + "(?, ?, ?, ?)")) {
+        ps.setBytes(1, customerId.toByteArray());
+        ps.setBytes(2, accountId.toByteArray());
+        ps.setBytes(3, roleId.toByteArray());
+        ps.setString(4, roleName);
+        ps.executeUpdate();
+        System.out.printf("New role created: %s\n", roleName);
+      }
+    }
   }
 
-  void moveAccountBalance(ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount) throws SpannerException {
-    databaseClient
-        .readWriteTransaction()
-        .run(
-            transaction -> {
-              // Get account balances.
-              ImmutableMap<ByteArray, BigDecimal> accountBalances =
-                  readAccountBalances(fromAccountId, toAccountId, transaction);
-
-              transaction.buffer(
-                  ImmutableList.of(
-                      buildUpdateAccountMutation(
-                          fromAccountId, accountBalances.get(fromAccountId).subtract(amount)),
-                      buildUpdateAccountMutation(
-                          toAccountId, accountBalances.get(toAccountId).add(amount)),
-                      buildInsertTransactionHistoryMutation(
-                          fromAccountId, amount, /* isCredit= */ true),
-                      buildInsertTransactionHistoryMutation(
-                          toAccountId, amount, /* isCredit= */ false)));
-              return null;
-            });
+
+  void moveAccountBalance(
+      ByteArray fromAccountId, ByteArray toAccountId, BigDecimal amount)
+      throws SQLException {
+    try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+      // begin transaction
+      connection.setAutoCommit(false);
+      try (
+          PreparedStatement readStatement = connection.prepareStatement(
```

Comment:
It seems to be common practice to instantiate `Statement`s and `PreparedStatement`s right after `Connection`s. However, these statements are passed later to helper functions to populate the `?` variables, distancing the code that modifies the SQL commands from the commands themselves. I added comments that contain the SQL commands to improve readability, but another option could be to pass the `connection` rather than `PreparedStatement`s.